components/script/dom/bindings/principals.rs

```diff
@@ -0,0 +1,122 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use js::{
+    glue::{
+        CreateRustJSPrincipals, DestroyRustJSPrincipals, GetRustJSPrincipalsPrivate,
+        JSPrincipalsCallbacks,
+    },
+    jsapi::{JSPrincipals, JS_DropPrincipals, JS_HoldPrincipals},
+    rust::Runtime,
+};
+use servo_url::MutableOrigin;
+use std::{marker::PhantomData, ops::Deref, ptr::NonNull};
+
+/// An owned reference to Servo's `JSPrincipals` instance.
+#[repr(transparent)]
+pub struct ServoJSPrincipals(NonNull<JSPrincipals>);
+
+impl ServoJSPrincipals {
+    pub fn new(origin: &MutableOrigin) -> Self {
+        unsafe {
+            let private: Box<MutableOrigin> = Box::new(origin.clone());
+            let raw = CreateRustJSPrincipals(&PRINCIPALS_CALLBACKS, Box::into_raw(private) as _);
+            // The created `JSPrincipals` object has an initial reference
+            // count of zero, so the following code will set it to one
+            Self::from_raw_nonnull(NonNull::new_unchecked(raw))
+        }
+    }
+
+    /// Construct `Self` from a raw `*mut JSPrincipals`, incrementing its
+    /// reference count.
+    #[inline]
+    pub unsafe fn from_raw_nonnull(raw: NonNull<JSPrincipals>) -> Self {
+        JS_HoldPrincipals(raw.as_ptr());
+        Self(raw)
+    }
+
+    #[inline]
+    pub unsafe fn origin(&self) -> MutableOrigin {
+        let origin = GetRustJSPrincipalsPrivate(self.0.as_ptr()) as *mut MutableOrigin;
+        (*origin).clone()
+    }
+
+    #[inline]
+    pub fn as_raw_nonnull(&self) -> NonNull<JSPrincipals> {
+        self.0
+    }
+
+    #[inline]
+    pub fn as_raw(&self) -> *mut JSPrincipals {
+        self.0.as_ptr()
+    }
+}
+
+impl Clone for ServoJSPrincipals {
+    #[inline]
+    fn clone(&self) -> Self {
+        unsafe { Self::from_raw_nonnull(self.as_raw_nonnull()) }
+    }
+}
+
+impl Drop for ServoJSPrincipals {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe { JS_DropPrincipals(Runtime::get(), self.as_raw()) };
+    }
+}
+
+/// A borrowed reference to Servo's `JSPrincipals` instance.
+#[derive(Clone, Copy)]
+pub struct ServoJSPrincipalsRef<'a>(NonNull<JSPrincipals>, PhantomData<&'a ()>);
+
+impl ServoJSPrincipalsRef<'_> {
+    /// Construct `Self` from a raw `NonNull<JSPrincipals>`.
+    #[inline]
+    pub unsafe fn from_raw_nonnull(raw: NonNull<JSPrincipals>) -> Self {
+        Self(raw, PhantomData)
+    }
+
+    /// Construct `Self` from a raw `*mut JSPrincipals`.
+    ///
+    /// # Safety
+    ///
+    /// The behavior is undefined if `raw` is null.
+    #[inline]
+    pub unsafe fn from_raw_unchecked(raw: *mut JSPrincipals) -> Self {
+        Self::from_raw_nonnull(NonNull::new_unchecked(raw))
+    }
+}
+
+impl Deref for ServoJSPrincipalsRef<'_> {
+    type Target = ServoJSPrincipals;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(&self.0 as *const NonNull<JSPrincipals> as *const ServoJSPrincipals) }
```

Comment:
Let's do `&*(self.0.as_ptr().cast::<ServoJSPrincipals>())` instead.

Reply:
That would change the code's semantics - it's intended to reinterpret `&self.0: &NonNull<JSPrincipals>` (not `self.0`) as `&ServoJSPrincipals` (i.e., `&NonNull<JSPrincipals>`).

Reply:
This approach lets us have `ServoJSPrincipalsRef: Copy`, but on second thought, this might not be worth the extra `unsafe` block. I implemented a different approach in commit e98cba1896160e6512d8f74988d253cd5b1047e4.